### PR TITLE
Refactor ByteSlice

### PR DESCRIPTION
**Pull request by Nugine**
replace `crate::fuse::byte_slice::ByteSlice` with `crate::fuse::de::Deserializer`

### CODE REVIEW

File: async_fuse/src/fuse/decoder.rs [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+            "no enough bytes to fetch, remaining {} bytes but to fetch {} bytes",
+            len,
+            need,
+        );
+        return Err(DecodeError::NotEnough);
+    }
+    Ok(())
+}
+
+impl<'b> Decoder<'b> {
```
Comment (rogercloud) [inline async_fuse/src/fuse/decoder.rs:80]: The functions in the struct Decoder is all about fetch data from u8 slice, while the decode function is actually defined in the T structure. So I suggest to rename this to something related to buf or slice, for example slice generator?

The current method call example is:
```
SomeTypeDeriveDecode a = decoder.decode();
```

I prefer the following one:
```
SomeTypeDeriveDecode a = SomeTypeDeriveDecode::decode(slice_generator);
```
It's more clear that the decode function is defined in the `SomeTypeDeriveDecode`, and the generator only handles the slice stuff.

File: async_fuse/src/fuse/decoder.rs [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+                trace!("no trailing zero in bytes, cannot fetch c-string");
+                DecodeError::NotEnough
+            })?
+            .wrapping_add(1);
+        debug_assert!(strlen <= self.bytes.len());
+        unsafe { Ok(self.pop_bytes_unchecked(strlen)) }
+    }
+
+    /// Fetch some bytes as a OS-string
+    #[cfg(any(target_os = "linux", target_os = "macos"))]
```
Comment (rogercloud) [inline async_fuse/src/fuse/decoder.rs:205]: We can ignore the macos curretnly.

File: async_fuse/src/fuse/decoder.rs [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+pub trait Decode<'b>: Sized {
+    /// Decode from bytes
+    fn decode(de: &'_ mut Decoder<'b>) -> Result<Self, DecodeError>;
+}
+
+/// The error returned by `Decoder`
+#[derive(Debug, PartialEq, Eq, thiserror::Error)]
+pub enum DecodeError {
+    /// Expected more data
+    #[error("NotEnough")]
```
[the remark this answers is not in the document]
Comment (Nugine) [reply async_fuse/src/fuse/decoder.rs:30]: `DecodeError` is an internal error type but not a report type. (`anyhow::Error` is a report type)
There is more information in the context and logs.

File: async_fuse/src/fuse/decoder.rs (hunk 1 above)
Comment (Nugine) [reply async_fuse/src/fuse/decoder.rs:80]: `Decode`/`Decoder` are just like `Deserialize`/`Deserializer` in `serde`.
<https://docs.rs/serde/1.0.118/serde/de/index.html>

byte_slice is not a slice though there is a slice in it.
slice_generator is not a [`Generator`](https://doc.rust-lang.org/stable/std/ops/trait.Generator.html).
I think both of them are misleading.

File: async_fuse/src/fuse/decoder.rs (hunk 2 above)
Comment (Nugine) [reply async_fuse/src/fuse/decoder.rs:205]: shall we emit a compile error when compiles on macos?
```rust
// src/lib.rs
#[cfg(target_os = "macos")]
compile_error!("async-fuse does not support macos")
```

File: async_fuse/src/fuse/decoder.rs (hunk 1 above)
Comment (Nugine) [reply async_fuse/src/fuse/decoder.rs:80]: How about making `fetch_*` free functions which take `&mut &[u8]` as input?

**Review by rogercloud — APPROVED** [review]

**Pull request merged**